### DISCUSSION TITLE
gtk-fortran as a dependency using the Fortran Package Manager (fpm)

### DISCUSSION THREAD

**Discussion by vmagnin** [vmagnin/gtk-fortran#246] [Ideas]
The new _Fortran Package Manager_ (fpm) can import dependencies from the web using git. For the moment it is limited to dependencies based on fpm. But in the future it should be able to import other projects which propose a build script, via the `build-script` variable of their fpm.toml file. Or even directly call CMake.

Our first attempt to use fpm is related here: https://github.com/fortran-lang/fpm/issues/351

See also: 
https://fortran-lang.discourse.group/t/google-summer-of-code-2021/658
https://github.com/fortran-lang/fortran-lang.org/wiki/GSoC-2021-Project-ideas

**Comment by Zaya6** [comment]
any progress on this?

**Reply by vmagnin** [reply]
If you clone the experimental_fpm branch of https://github.com/vmagnin/gtkzero_fpm
and do fpm run, it should clone gtk-fortran as a dependency, and launch an empty GTK window. You will need of course to install the GTK4 dev files.

Before closing this discussion, I will finalize the work and test it on several platforms, then merge in the main gtk4 branch.

**Comment by vmagnin** [comment]
The Fortran-lang community is working hard on fpm, but we must be patient. In order to use gtk-fortran as an fpm dependency, we need at least those two features:

- fpm must be able to manage CMake based projects: https://github.com/fortran-lang/fpm/issues/69
    - It is a part of the ideas proposed for the 2021 Google Summer of Code: https://github.com/fortran-lang/fortran-lang.org/wiki/GSoC-2021-Project-ideas#source-code-distribution-with-foreign-build-systems
 - fpm must be able to manage included files (`gtkenums-auto.f90` and `gtk-auto.f90`): https://github.com/fortran-lang/fpm/issues/358
 
Note that there is already many candidates for the GSoC 2021: https://fortran-lang.discourse.group/t/google-summer-of-code-2021/658/50

**Comment by vmagnin** [comment]
I am making good progress in using gtk-fortran as a fpm dependency:
https://github.com/fortran-lang/fpm/issues/351#issuecomment-1092656731

Finally, we don't need fpm to manage CMake for that. And the included files problem is now fixed in gtk-4-fortran. I hope to finish working on that feature for gtk-fortran 4.2, which could be released in May 2022 and will be based on GTK 4.6.

**Comment by vmagnin** [comment]
The two experimental_fpm branches have been merged into the gtk4 and main branches of the two projects:
https://github.com/vmagnin/gtk-fortran
https://github.com/vmagnin/gtkzero_fpm
That killing feature will be announced with the release of gtk-fortran 4.2 in a few weeks.

But before, in the experimental_fpm branches I will now explore the use of PLplot when gtk-fortran is used as a fpm dependency.

**Reply by vmagnin** [reply]
The `experimental_fpm` branch has been merged into gtk4 : PLplot can be used when gtk-fortran is used as a fpm dependency.

**Comment by vmagnin** [comment]
gtk-fortran can now be used as a fpm dependency:
see https://github.com/vmagnin/gtk-fortran/releases/tag/v4.2.0
